### ISSUE TITLE
fread not reading quoted columns as character

### ISSUE THREAD

**Issue by hutch3232** [Rdatatable/data.table#3756]
With older versions of data.table, I could convert variables that _look_ like numeric fields into character. Then I would _fwrite_ them, with the goal being to later _fread_ them back with the data types preserved.

Here is a really simple data.table with the first three columns being character and the last integer.
```
library(data.table)

my_data <- data.table(
  var1 = c("001", "002", "003", "004", "005"),
  var2 = c("1", "2", "3", "4", "5"),
  var3 = c("a", "b", "c", "d", "e"),
  var4 = 1:5
)

str(my_data)

Classes ‘data.table’ and 'data.frame':	5 obs. of  4 variables:
 $ var1: chr  "001" "002" "003" "004" ...
 $ var2: chr  "1" "2" "3" "4" ...
 $ var3: chr  "a" "b" "c" "d" ...
 $ var4: int  1 2 3 4 5
 - attr(*, ".internal.selfref")=<externalptr> 
```
I want to be sure the data is saved with the proper formatting, so when I run _fwrite_ I include the arguments _quotes_ and _qmethod_ like this:

```
fwrite(x = my_data, file = "my_data.csv", quote = TRUE, qmethod = "double")
```
I now _fread_ the data back in and inspect the structure. Unfortunately, var1 and var2 are now integer when they should be character.
```
my_data_attempt1 <- fread(file = "my_data.csv")
str(my_data_attempt1)

Classes ‘data.table’ and 'data.frame':	5 obs. of  4 variables:
 $ var1: int  1 2 3 4 5
 $ var2: int  1 2 3 4 5
 $ var3: chr  "a" "b" "c" "d" ...
 $ var4: int  1 2 3 4 5
 - attr(*, ".internal.selfref")=<externalptr> 
```
I have found that one specific argument, _keepLeadingZeros_, does work to solve the issue for var1, but var2 is coerced.
```
my_data_attempt2 <- fread(file = "my_data.csv", keepLeadingZeros = TRUE)
str(my_data_attempt2)

Classes ‘data.table’ and 'data.frame':	5 obs. of  4 variables:
 $ var1: chr  "001" "002" "003" "004" ...
 $ var2: int  1 2 3 4 5
 $ var3: chr  "a" "b" "c" "d" ...
 $ var4: int  1 2 3 4 5
 - attr(*, ".internal.selfref")=<externalptr> 
```
I'm showing this example on 1.12.3, but I'm also having the issue on 1.12.2. 
```
> sessionInfo()
R version 3.6.0 (2019-04-26)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18362)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252 LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.12.3
```

**Comment by franknarf1** [comment]
Just looking at `args(fread)` and `args(fwrite)`, I seem to have found the current way to do it:

```
> fwrite(my_data)
var1,var2,var3,var4
001,1,a,1
002,2,b,2
003,3,c,3
004,4,d,4
005,5,e,5
> fwrite(my_data, quote=TRUE)
"var1","var2","var3","var4"
"001","1","a",1
"002","2","b",2
"003","3","c",3
"004","4","d",4
"005","5","e",5
> out = capture.output(fwrite(my_data, quote=TRUE))
> fread(paste(out, collapse="\n"), keepLeadingZeros = TRUE)
   var1 var2 var3 var4
1:  001    1    a    1
2:  002    2    b    2
3:  003    3    c    3
4:  004    4    d    4
5:  005    5    e    5
```

I assume that if default args worked for you before, the change was noted in the NEWS file:  https://github.com/Rdatatable/data.table/blob/master/NEWS.md

**Comment by jangorecki** [comment]
I don't have any numbers to back that up but I think that more users would expect quoted numeric column to be read as numeric, so I would call that feature. For now you can try `colClasses` argument.

**Comment by hutch3232** [comment]
I agree with you that most people would expect quoted numerics to be read in as numeric. I suspect that is why the default method of _fwrite_ is to write out the numeric columns without quotes (I can see this when opening the .csv in notepad++). 
```
fwrite(x = my_data, file = "my_data.csv")

var1,var2,var3,var4
001,1,a,1
002,2,b,2
003,3,c,3
004,4,d,4
005,5,e,5

# as opposed to my explicit quoting approach

fwrite(x = my_data, file = "my_data.csv", quote = TRUE, qmethod = "double")

"var1","var2","var3","var4"
"001","1","a",1
"002","2","b",2
"003","3","c",3
"004","4","d",4
"005","5","e",5
```
I think the difference here is that I'm explicitly quoting the numerics so that _fread_ will know when it's read in that I want those specific columns to remain character.

**Comment by hutch3232** [comment]
Following up on this issue, it seems this change was implemented in release 1.11.0:

> Numeric data that has been quoted is now detected and read as numeric.

While I still think that if data is quoted in the .csv, it is intended to be read in as character, since this has been the new behavior for a while, perhaps an alternative could be implemented. A new `fread` argument could be written, say, `quotedNumericsAsCharacter` with the default being `FALSE`. Then users could choose if those columns should be coerced to numeric, rather than it being automatic.

**Comment by jflycn** [comment]
This is really really bad. I thought adding "quote = TRUE" to fwrite can make sure the columns can be read back as characters. 

Since numeric data will always be read as numeric, no matter quoted or not, what is the point of adding "quote = TRUE" to fwrite then?

Maybe "quote = TRUE" should be removed, so that it will not create more confusion?

Also, in the documentation of the `fread` function, under "Quotes":

> Quoting may be used to signify that numeric data should be read as text. 

Should you change it to something like: "Quoting may be used to signify that numeric data should be read as text, but in fact we will still convert them to integer or numeric. "

**Comment by jflycn** [comment]
Also, for those who suddenly found fread no longer works as expected because of this changing of behavior, one solution might be adding `yaml = TRUE` to both of `fwrite` and `fread`. 

You can use an editor like notepad++ to find all the `fwrite` and `fread` in your code base. It might be tedious to add hundreds of `yaml = TRUE`, but it is certainly easier then using the `colClasses ` argument, if you have different columns need to handle in different data files.

`keepLeadingZeros = TRUE` can partially solve the problem, but be careful when you have a filtered data set which happens to having no leading zeros.

If you do use `yaml = TRUE`, be careful not to use it on external data files.
